red_black_tree/rb_tree.py

```
@@ -0,0 +1,552 @@
+        return bool(self.get_children_count())
+
+    def get_children_count(self):
+        """ Returns the number of NOT NIL children the node has """
+        if self.color == NIL:
+            return 0
+        return sum([int(self.left.color != NIL), int(self.right.color != NIL)])
+
+
+class RedBlackTree:
+    # every node has null nodes as children initially, create one such object
+    # for easy management
+    NIL_LEAF = Node(value=None, color=NIL, parent=None)
+
+    def __init__(self):
+        self.count = 0
+        self.root = None
+        self.ROTATIONS = {
+            # Used for deletion and uses the sibling's relationship with his
+            # parent as a guide to the rotation
+            'L': self._right_rotation,
+            'R': self._left_rotation
+        }
+
+    def __iter__(self):
+        if self.root:
+            for r in self.root.__iter__():
+                yield r
+        else:
+            yield list()
+
+    def add(self, value):
+        if not self.root:
+            self.root = Node(value, color=BLACK, parent=None,
+                             left=self.NIL_LEAF, right=self.NIL_LEAF)
+            self.count += 1
+            return
+        parent, node_dir = self._find_parent(value)
+        if node_dir is None:
+            return  # value is in the tree
+        new_node = Node(value=value, color=RED, parent=parent,
+                        left=self.NIL_LEAF, right=self.NIL_LEAF)
+        if node_dir == 'L':
+            parent.left = new_node
+        else:
+            parent.right = new_node
+
+        self._try_rebalance(new_node)
+        self.count += 1
+
+    def remove(self, value):
+        """
+        Try to get a node with 0 or 1 children.
+        Either the node we're given has 0 or 1 children or we get its successor.
+        """
+        node_to_remove = self.find_node(value)
+        if node_to_remove is None:  # node is not in the tree
+            return
```

This should raise an exception here. maybe `ValueError`